### PR TITLE
Firestore: Fix global refs exhaustion issue in Android

### PR DESCRIPTION
**Pull request by dconeybe**
Properly fix the the global refs exhaustion issue tracked by https://github.com/firebase/firebase-unity-sdk/issues/569, and reported in https://github.com/firebase/quickstart-unity/issues/1303 and https://github.com/firebase/quickstart-unity/issues/1193#issuecomment-1242203289 (Googlers see b/251869890 for more details). Part 1, a partial mitigation, was https://github.com/firebase/firebase-cpp-sdk/pull/1111.

This PR supersedes, and builds upon the work done by @cherylEnkidu in, https://github.com/firebase/firebase-cpp-sdk/pull/1176.

### CODE REVIEW

**Review by wu-hui — APPROVED** [review]
This is an impressive PR, I love how organized it is.

I left couple comments on potential improvements in some code comments, but LGTM overall, nice work!

File: firestore/src_java/com/google/firebase/firestore/internal/cpp/ObjectArena.java [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+ * limitations under the License.
+ */
+
+package com.google.firebase.firestore.internal.cpp;
+
+import java.util.HashMap;
+
+/**
+ * A companion class used by the C++ class firebase::firestore::jni::ArenaRef to
+ * map long primitive values to Objects.
```
[the remark this answers is not in the document]
Comment (dconeybe) [reply firestore/src_java/com/google/firebase/firestore/internal/cpp/ObjectArena.java]: Done.

File: firestore/src_java/com/google/firebase/firestore/internal/cpp/ObjectArena.java [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+ *
+ * Technically, this could have been written using JNI calling HashMap methods
```
[the remark this answers is not in the document]
Comment (dconeybe) [reply firestore/src_java/com/google/firebase/firestore/internal/cpp/ObjectArena.java]: Done.

File: firestore/src/jni/arena_ref.h [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+   *
+   * Subsequent invocations of `get()` will return the given object. The given
+   * object may be a Java "null" object.
+   *
+   * If invoked with a pending Java exception then this `ArenaRef` is set to a
+   * `null` Java object reference.
+   */
+  void reset(Env& env, const Object&);
+
+  static void Initialize(Loader&);
```
[the remark this answers is not in the document]
Comment (dconeybe) [reply firestore/src/jni/arena_ref.h:114]: Comments added.

**Pull request merged**